### ISSUE TITLE
Array-as-vec performance regression from #12867

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#12895]
#12867 added denormalization to condstmts, but appear to have introduced some performance regression for push_back and push_front -- [link](https://chapel-lang.org/perf/chapcs/?startdate=2019/04/11&enddate=2019/04/25&graphs=arrayvectoroperations):

![image](https://user-images.githubusercontent.com/1588337/56775274-2cb1d100-677b-11e9-8219-0d6680e4bb3c.png)

Here's a smaller reproducer:

```chpl
use Time;

config const n: int = 50000000;
config const printTimes = true;

class Runner {
  proc run(A) { halt("virtual function run called"); }
}

class PushBack: Runner {
  const n: int;
  override proc run(A: [] int) {
    for i in 1..n {
      A.push_back(i);
    }
  }
}

proc main {
  var A: [1..0] int;
  var r: owned Runner;

  r = new owned PushBack(n);
  var t: Timer; t.start();
  r.run(A);
  writeln(t.elapsed());
}
```

If you remove the virtual dispatch (just do `var  r = new owned PushBack(n))` performance seems better, so this is probably a side-effect and nothing that serious, but it'd be good to understand the root cause.

**Comment by vasslitvinov** [comment]
We fall prey to numeric instability here, just like with did with mandelbrot. (OK, not quite that, yet similarly surprising.)

Applying this patch to the generated code for Elliot's reproducer above:

```diff
--- ChapelArray.c
+++ ChapelArray.c
@@ -390,7 +390,8 @@
   /*  969127 */ BaseDist call_tmp69 = NULL;
   /* 1312325 */ chpl_bool call_tmp70;
   /* 1312345 */ chpl_bool call_tmp71;
-  if (! (this11)->_unowned) {
+  chpl_bool call_tmp72 = (! (this11)->_unowned);
+  if (call_tmp72) {
     /*  969255 */ coerce_tmp = (this11)->_instance;
     /*  969070 */ remove3(((BaseDom)(coerce_tmp)), &ret_tmp, _ln, _fn);
     /* 1193983 */ call_tmp67 = ret_tmp;
```

brings the timed region from 0.45s to 0.37s. This is the difference we observe in the performance regression sited above.

Here is [the full text of the function](https://gist.github.com/vasslitvinov/d4d6dcf57543d3df9270976dc4e7130c) where the above patch applies.

**Comment by bradcray** [comment]
The `+` lines result in the 0.37s version and the `-` lines to the 0.45s version?

**Comment by vasslitvinov** [comment]
> The `+` lines result in the 0.37s version and the `-` lines to the 0.45s version?

Correct. `+`-lines show the code without denormalization. It runs faster.

**Comment by bradcray** [comment]
Weird.  What if a cast to chpl_bool is inserted into the `-` version?

**Comment by vasslitvinov** [comment]
I tried `if ((chpl_bool)(! (this11)->_unowned))`

That did not make a difference.
